File: cpp/arrays/reverse.cpp
#include <bits/stdc++.h>
using namespace std;

void swap(int arr[], int start, int end){
    arr[start] = arr[start] ^ arr[end];
    arr[end] = arr[start] ^ arr[end];
    arr[start] = arr[start] ^ arr[end];
}
void printArray(int arr[], int n){
    for(int i=0; i<n; i++)
        cout << arr[i] << " ";
}

int reverseIterative(int arr[], int start, int end, int n){
    while(start < end) {
        swap(arr, start, end);
        start++;
        end--;
    }
}

int reverseRecursive(int arr[], int start, int end){
    if(start >= end)
        return NULL;
    swap(arr, start, end);
}

int main(){
    int arr[] = {1, 4, 6, 4, 6, 2, 7};
    int n = sizeof(arr)/ sizeof(arr[0]);

    cout << "Array: ";
    printArray(arr, n);
    cout << "\nReversed array: ";
//    reverseIterative(arr, 0, n-1, n);
    reverseRecursive(arr, 0, n-1);
    printArray(arr, n);
}

File: Arrays/reverse.py
def reverseIterative(arr):
    n = len(arr)
    for i in range(n//2):
        arr[i], arr[n-i-1] = arr[n-i-1], arr[i]
    return arr

def reverseRecursive(arr, start, end):
    if start >= end:
        return
    arr[start], arr[end] = arr[end], arr[start]
    reverseRecursive(arr, start+1, end-1)
    return arr

def reverseList(arr):
    return arr[::-1]

arr = [1, 2, 3, 4, 5, 6, 7]
print("Array: {}".format(arr))

# revArray = reverseIterative(arr)
revArray = reverseRecursive(arr, 0, len(arr)-1)
# revArray = reverseList(arr)

print("Reversed array: {}".format(revArray))